tests/cases/unittests/session.ts

```diff
@@ -0,0 +1,459 @@
+		readFile: () => void 0,
+		writeFile: () => void 0,
+		resolvePath: () => void 0,
+		fileExists: () => false,
+		directoryExists: () => false,
+		createDirectory: () => void 0,
+		getExecutingFilePath: () => void 0,
+		getCurrentDirectory: () => void 0,
+		readDirectory: () => void 0,
+		exit: () => void 0
+	},
+	mockLogger: Logger = {
+        close(): void {},
+        isVerbose(): boolean { return false; },
+        loggingEnabled(): boolean { return false; },
+        perftrc(s: string): void {},
+        info(s: string): void {},
+        startGroup(): void {},
+        endGroup(): void {},
+        msg(s: string, type?: string): void {},
+	};
+	
+	describe('the Session class', () => {
+		let session: Session,
+			lastSent: protocol.Message;
+
+		beforeEach(() => {
+			session = new Session(mockHost, Buffer.byteLength, process.hrtime, mockLogger);
+			session.send = (msg: protocol.Message) => {
+				lastSent = msg;
+			};
+		});
+
+		describe('executeCommand', () => {
+			it('should throw when commands are executed with invalid arguments', () => {
+				let req : protocol.FileRequest = {
+					command: CommandNames.Open,
+					seq: 0,
+					type: 'command',
+					arguments: {
+						file: undefined
+					}
+				};
+
+				expect(() => session.executeCommand(req)).to.throw();
+			});
+			it('should output an error response when a command does not exist', () => {
+				let req : protocol.Request = {
+					command: 'foobar',
+					seq: 0,
+					type: 'command'
+				};
+
+				session.executeCommand(req);
+
+				expect(lastSent).to.deep.equal(<protocol.Response>{
+					command: CommandNames.Unknown,
+					type: 'response',
+					seq: 0,
+					message: 'Unrecognized JSON command: foobar',
+					request_seq: 0,
+					success: false
+				});
+			});
+			it('should return a tuple containing the response and if a response is required on success', () => {
+				let req : protocol.ConfigureRequest = {
+					command: CommandNames.Configure,
+					seq: 0,
+					type: 'command',
+					arguments: {
+						hostInfo: 'unit test',
+						formatOptions: {
+							newLineCharacter: '`n'
+						}
+					}
+				};
+
+				expect(session.executeCommand(req)).to.deep.equal({
+					responseRequired: false
+				});
+				expect(lastSent).to.deep.equal({
+					command: CommandNames.Configure,
+					type: 'response',
+					success: true,
+					request_seq: 0,
+					seq: 0,
+					body: undefined
+				});
+			});
+		});
+
+		describe('onMessage', () => {
+			it('should not throw when commands are executed with invalid arguments', () => {
+				let i = 0;
+				for (name in CommandNames) {
+					if (!Object.prototype.hasOwnProperty.call(CommandNames, name)) {
+						continue;
+					}
+					let req : protocol.Request = {
+						command: name,
+						seq: i++,
+						type: 'command'
+					};
+					session.onMessage(JSON.stringify(req));
+					req.seq += 2;
+					req.arguments = {};
+					session.onMessage(JSON.stringify(req));
+					req.seq += 2;
+					req.arguments = null;

```

Comment:
I'm not necessarily checking for expected values here, I'm trying to check that the service doesn't throw when you pass invalid things (or anything) to it 'over the wire'. I should probably actually try passing both (And I think I do - the first omits `arguments` which is equivalent to setting it to `undefined` afaik?)

I should also probably add a test for passing garbage strings to onMessage, rather than just json with invalid/empty arguments.